src/applications/merkle_tree.md

```diff
@@ -0,0 +1,48 @@
+# Merkle Tree contract
+
+A Merkle tree, also known as a hash tree, is a data structure used in cryptography and computer science to verify data integrity and consistency. It is a binary tree where each leaf node represents the cryptographic hash of some data (a transaction for example), and each non-leaf node represents the cryptographic hash of its child nodes. This hierarchical structure allows efficient and secure verification of the data integrity.
+
+Here's a quick summary of how it operates and what functionalities it supports:
+
+### How it works:
+
+1. Leaves Creation:
+    - Some data is hashed to create a leaf node.
+2. Intermediate Nodes Creation:
+    - Pairwise hashes of the leaf nodes are combined and hashed again to create parent nodes.
+    - This process continues until only one hash remains, known as the Merkle root.
+3. Merkle Root:
+    - The final hash at the top of the tree, representing the entire dataset.
+    - Changing any single data block will change its corresponding leaf node, which will propagate up the tree, altering the Merkle root.
+
+### Key Features:
+
+1. Efficient Verification:
+    - Only a small subset of the tree (the Merkle proof) is needed to verify the inclusion of a particular data block, reducing the amount of data that must be processed.
+
+2. Data Integrity:
+    - The Merkle root ensures the integrity of all the underlying data blocks.
+    - Any alteration in the data will result in a different root hash.
+
```

Comment:
Could you extend and show some valuable/common use cases in blockchain? By keeping the merkle root on-chain, and keep the data in an off-chain storage (example: whitelist inclusion with merkle proof)